pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -109,9 +109,48 @@ func (p *staticPolicy) Name() string {
 }
 
 func (p *staticPolicy) Start(s state.State) {
-	// Configure the shared pool to include all detected CPU IDs.
-	allCPUs := p.topology.CPUDetails.CPUs()
-	s.SetDefaultCPUSet(allCPUs)
+	// flag to indicate whether ignore state from file and start fresh state
+	initializeNewEmptyState := false
+
+	// static policy must have at least one cpu in defaultCPUSet
+	if s.GetDefaultCPUSet().IsEmpty() {
+		// State is initialized empty
+		// Configure the shared pool to include all detected CPU IDs.
+		initializeNewEmptyState = true
+	} else {
+		// State has already been initialized
+
+		// 1 Reserved cpuset is not part of default cpuset because:
+		// - kube/system reserved have changed (increased) - may lead to some containers not being able to start
+		// - user tampered with file
+		if !p.reserved.Intersection(s.GetDefaultCPUSet()).Equals(p.reserved) {
+			glog.Warningf("[cpumanager] static policy: Start - not all reserved cpus: \"%s\" are present in defaultCpuSet: \"%s\"",
+				p.reserved.String(), s.GetDefaultCPUSet().String())
+			initializeNewEmptyState = true
+		}
+
+		// 2. Check if state for static policy is consistent
+		if !initializeNewEmptyState {
+			tmpAssignments := s.GetCPUAssignments()
+			tmpDefaultCPUset := s.GetDefaultCPUSet()
+			for cID, cset := range tmpAssignments {
+				// Non of the cpu in DEFAULT cset should be in s.assignments
+				if !tmpDefaultCPUset.Intersection(cset).IsEmpty() {
+					initializeNewEmptyState = true
+					glog.Warningf("[cpumanager] static policy: Start (container id: %s cpuset: \"%s\") overlaps with default cpuset \"%s\"",
+						cID, cset.String(), tmpDefaultCPUset.String())
+					break
+				}
+			}
+		}
+	}
+
+	if initializeNewEmptyState {
```

Comment:
What do you think about this:

If I chose the static CPU manager I would like a pretty strong guarantee that assignments won’t change without operator notification.

One option, implemented here, is to log and start the state from scratch.

Another route could be to panic and tell the operator to drain the node amd delete the cpumanager state from disk.

Just putting this here for consideration. We can change our minds later but it’s worth thinking about. CC @jeremyeder @derekwaynecarr

Reply:
I prefer the second option (i.e., panic and ask the operator to drain the node). We should use this option we cannot re-assign the cpus in the state file (e.g., when kube- or system-reserved is changed) or when the state file is corrupted.

Reply:
@balajismaniam decreasing kube/system reserved is case we can handle , increasing on the other hand not so much - some pods may not be able to start.

Reply:
@flyingcougar Right. Hence I prefer "panic and ask the operator to drain the node" option as we don't have a clean solution to mitigate these issues.